[docs] Create newsletter_september2023.md (#732)
chanzuckerberg/cellxgene-census: docs/articles/2023/20231012-normalized_layer_precalc_stats.md

# Introducing a normalized layer and pre-calculated cell and gene statistics in Census

*Published: October 12, 2023*

*By: [Maximilian Lombardo](mailto:mlombardo@chanzuckerberg.com) and [Pablo Garcia-Nieto](pgarcia-nieto@chanzuckerberg.com)*

> The Census team is thrilled to announce the introduction of a library-size normalized expression layer and pre-calculated cell and gene statistics. This work is reflected in the Census schema V1.1.0 tailored to empower your single-cell research.

Reviewer: add link to schema

> The Census team is thrilled to announce the introduction of a library-size normalized expression layer and pre-calculated cell and gene statistics. This work is reflected in the Census schema V1.1.0 tailored to empower your single-cell research.

Reviewer: style: don't love the hyperbole of "thrilled".

> The Census team is happy to announce the introduction of a library-size normalized expression layer and pre-calculated cell and gene statistics. This work is reflected in changes introduced in the [Census schema V1.1.0](https://github.com/chanzuckerberg/cellxgene-census/blob/main/docs/cellxgene_census_schema.md) tailored to empower your single-cell research.

Reviewer 2: ```suggestion
The Census team is happy to announce the introduction of two new data features, tailored to empower your single-cell research: a library-size, normalized expression layer and pre-calculated cell and gene statistics. This work is reflected in changes introduced in the [Census schema V1.1.0](https://github.com/chanzuckerberg/cellxgene-census/blob/main/docs/cellxgene_census_schema.md)```

Reviewer 2: Also, what does "library-size" mean? Full Census?

Reviewer 3: sum of counts in cell, it's common single-cell jargon

> - Expand their Census query filters to select genes or cells based on the new metadata included, for example selecting only cells with N number of genes expressed.

Reviewer: remove "included"

> - Expand their Census query filters to select genes or cells based on the new metadata, for example selecting only cells with N number of genes expressed.

Reviewer 2: ```suggestion
- Expand their Census query filters to select genes or cells based on the new metadata. For example, selecting only cells with N number of genes expressed.
```

> These features are currently exclusive to the "latest" versions of the Census data release. We invite your feedback as you explore these novel functionalities.

Reviewer: perhaps note they will be included in the next LTS?

Keep on reading to find out more about these features!

## Description of new data added to Census

All of the following changes were introduced in the Census schema V1.1.0

### Added a new library-size normalized layer

> We have introduced a library-size normalized X layer for the RNA measurements of both the human and mouse experiments under `X["normalized"]`.  The normalized layer is built by dividing each value in the raw count matrix by its corresponding row sum (i.e. size normalization).

Reviewer: "under" is awkward.  Perhaps "available as X['normalized']" or "saved as..."?

> It is important to note that in order to preserve high precision for the result of a division we would need to encode data with many decimal point values. This can lead to a representation of data that is very large on disk and on memory. As means to compress this matrix we introduced changes that lead to reduced numerical precision and slight deviations from the expected row sum of 1.0.

Reviewer: this paragraph is super confusing (accurate, but hard to parse).

Perhaps something like: 
To reduce data size and improve performance, normalized values are stored with a reduced floating point precision. In addition, to ensure that small count values do not round to zero, a small sigma has been added. You will see the effect of these artifacts in row (per-cell) values not summing to precisely 1.0.

To reduce data size and improve performance, normalized values are stored with a reduced floating point precision. In addition, to ensure that small count values do not round to zero, a small sigma has been added. You will see the effect of these artifacts in row (per-cell) values not summing to precisely 1.0.

### Enhanced gene metadata

> The `ms["RNA"].var` data frame for both the human and mouse experiments has been enriched with two new metadata fields:

Reviewer 2: ```suggestion
The `ms["RNA"].var` DataFrame for both the human and mouse experiments has been enriched with two new metadata fields:
```

> - `nnz` the number of non-zero (nnz) values, effectively the number of cells expressing this gene.

Reviewer: the list needs punctuation between the field and field description, e.g., "nnz: the number ..."

Reviewer: oh, I notice it exists in a few places, and is just not consistently used.

> - `nnz` the number of non-zero (nnz) values, effectively the number of cells expressing this gene.

Reviewer: IIRC, this is wrong -- `nzz` is the number of explicitly stored values, not the number of non-zero values.

Reviewer 3: I can correct the definition, but the sentence is "effectively the number of cells expressing this gene" is still correct, as Census does not store explicit store 0s, right?

Reviewer: Yes.  But given how SOMA defines nnz, I thought it worth being explicit.  I leave it to you to decide :-)

- `n_measured_obs` — the "measured" cells for this gene, effectively the number of cells for which this gene was measured in their respective dataset.

### Enhanced cell metadata

> The `obs` data frame for both the human and mouse experiments is now augmented with the following new metadata, allowing users to forego common calculations used in early data pre-processing. For each cell:

Reviewer 2: ```suggestion
The `obs` DataFrame for both the human and mouse experiments is now augmented with the following new metadata, allowing users to forego common calculations used in early data pre-processing. For each cell:
```

> - `raw_sum` — the count sum derived from `X["raw"]`.

Reviewer 2: ```suggestion
- `raw_sum` — the sum of the raw counts, derived from `X["raw"]`.
```

Reviewer 2: I find this rewording clearer, but YMMV

- `nnz`: the number of non-zero (nnz) values, effectively the number of genes expressed on this cell.

> - `raw_mean` the average counts.

Reviewer: these names appear to be wrong - they should be raw_mean_nnz, etc.

Reviewer: And given the subtlety of the denominator (ie.. mean of explicitly stored values), it may be worth calling that out, or suggesting "for more details, see the schema"

Reviewer 3: vestigial, this article started back in August. Will correct

> - `raw_variance` the variance of the counts.

Reviewer: name is wrong - missing _nnz

> - `n_measured_vars` — the "measured" genes, effectively the number of genes measured in the dataset from which the cell was originated from, thus all cells from the same dataset have the same value for this variable.

Reviewer 2: ```suggestion
- `n_measured_vars` — the "measured" genes, effectively the number of genes measured in the dataset from which the cell originated, thus all cells from the same dataset have the same value for this variable.
```

## How to use the new features

### Exporting the normalized data to existing single-cell toolkits

In Python, the normalized data can be exported into AnnData specifying the `X_name = "normalized"` argument of the `cellxgene.get_anndata()` method.

```python
import cellxgene_census

with cellxgene_census.open_soma(census_version = "latest") as census
    adata = cellxgene_census.get_anndata
        census = census,
        organism = "Homo sapiens",
        var_value_filter = "feature_id in ['ENSG00000161798', 'ENSG00000188229']",
        obs_value_filter = "cell_type == 'sympathetic neuron'",
        column_names = {"obs": ["tissue", "cell_type"]},

>         X_name = "normalized" # Specificy the normalized layer for this query

Reviewer 2: ```suggestion
        X_name = "normalized" # Specify the normalized layer for this query
```

    )
```

> Similarly, in R we can export the data to Seurat or SingleCellExperiment objects with the argument `X_layers` of the functions `get_seurat()` and `get_singel_cell_experiment()`.

Reviewer 2: ```suggestion
Similarly, in R we can export the data to Seurat or SingleCellExperiment objects with the argument `X_layers` of the functions `get_seurat()` and `get_single_cell_experiment()`.
```

   var_value_filter = gene_filter,
   obs_value_filter = cell_filter,
   obs_column_names = cell_columns,
   X_layers = layers
)

#Single Cell Experiment

library("SingleCellExperiment")

sce_obj <- get_single_cell_experiment(
   census = census,
   organism = organism,
   var_value_filter = gene_filter,
   obs_value_filter = cell_filter,
   obs_column_names = cell_columns,
   X_layers = layers
)

```

### Accessing library-size normalized data layer via TileDB-SOMA

For memory-efficient data retrieval, you can use TileDB-SOMA as outlined below. In Python this looks like the following.

```python

import cellxgene_census
import tiledbsoma

# Open context manager
with cellxgene_census.open_soma(census_version = "latest") as census:

    # Access human SOMA object
    human = census["census_data"]["homo_sapiens"]

    query = human.axis_query(
       measurement_name = "RNA",
       obs_query = tiledbsoma.AxisQuery(
           value_filter = "tissue == 'brain' and sex == 'male'"
       )
    )

    # Set iterable for normalized matrix
    iterator = query.X("normalized").tables()
    
    # Iterate over the normalized matrix.
    # Get an iterative slice as pyarrow.Table

>     raw_slice = next (iterator)

Reviewer 2: ```suggestion
    raw_slice = next(iterator)
```

    
    # Perform analysis
    
    # close the query
    query.close()
```

And the equivalent code in R.

```r
library("cellxgene.census")
library("tiledbsoma")

human <-  census$get("census_data")$get("homo_sapiens")
query <-  human$axis_query(
  measurement_name = "RNA",
  obs_query = SOMAAxisQuery$new(
    value_filter = "tissue == 'brain' & sex == 'male'"
  )
)

# Set iterable for normalized matrix
iterator <-  query$X("normalized")$tables()

# Iterate over the normalized matrix.
# Get an iterative slice as an Arrow Table
raw_slice <-  iterator$read_next()

# Perform analysis
```

### Utilizing pre-calculated stats for querying `obs` and `var`

To filter cells or genes based on pre-calculated statistics and export to AnnData, you can use the new metadata variables as value filters.

> For example you can you can add a filter to query cells with more than 500 genes expressed in addition to other filters. In Python this looks like the following.

Reviewer 2: ```suggestion
For example, you can add a filter to query cells with more than 500 genes expressed, along with other filters. In Python this looks like the following.
```

        column_names = {"obs": ["tissue", "cell_type"]},
        var_value_filter = "feature_id in ['ENSG00000161798', 'ENSG00000188229']",
    )

    print(adata)
```

In R, the equivalent code looks as follows.

```r
#Seurat
library("cellxgene.census")
library("Seurat")

census <- open_soma(census_version = "latest")

organism <- "Homo sapiens"
gene_filter <- "feature_id %in% c('ENSG00000107317', 'ENSG00000106034')"
cell_filter <-  "nnz > 500 & cell_type == 'sympathetic neuron'"
cell_columns <- c("tissue", "cell_type")

seurat_obj <- get_seurat(
   census = census,
   organism = organism,
   var_value_filter = gene_filter,
   obs_value_filter = cell_filter,
   obs_column_names = cell_columns
)

#Single Cell Experiment

library("SingleCellExperiment")

sce_obj <- get_single_cell_experiment(
   census = census,
   organism = organism,
   var_value_filter = gene_filter,
   obs_value_filter = cell_filter,
   obs_column_names = cell_columns
)
```

## Help us improve these data additions

We encourage you to engage with these new features in the Census API and share your feedback. This input is invaluable for the ongoing enhancement of the Census project.

> For further information on the new library-size normalized layer, please reach out to us at [soma@chanzuckerberg.com](soma@chanzuckerberg.com). To report issues or for additional feedback, refer to our [Census GitHub repository](https://github.com/chanzuckerberg/cellxgene-census/issues).

Reviewer: what about "further information" on the other features?   Suggest:   "For further information on any new feature, please reach out ...."

        column_names = {"obs": ["tissue", "cell_type"]},
        var_value_filter = "feature_id in ['ENSG00000161798', 'ENSG00000188229']",
    )

    print(adata)
```

In R, the equivalent code looks as follows.

```r
#Seurat
library("cellxgene.census")
library("Seurat")

census <- open_soma(census_version = "latest")

organism <- "Homo sapiens"
gene_filter <- "feature_id %in% c('ENSG00000107317', 'ENSG00000106034')"
cell_filter <-  "nnz > 500 & cell_type == 'sympathetic neuron'"
cell_columns <- c("tissue", "cell_type")

seurat_obj <- get_seurat(
   census = census,
   organism = organism,
   var_value_filter = gene_filter,
   obs_value_filter = cell_filter,
   obs_column_names = cell_columns
)

#Single Cell Experiment

library("SingleCellExperiment")

sce_obj <- get_single_cell_experiment(
   census = census,
   organism = organism,
   var_value_filter = gene_filter,
   obs_value_filter = cell_filter,
   obs_column_names = cell_columns
)
```

## Help us improve these data additions

We encourage you to engage with these new features in the Census API and share your feedback. This input is invaluable for the ongoing enhancement of the Census project.

> For further information on the new library-size normalized layer, please reach out to us at [soma@chanzuckerberg.com](soma@chanzuckerberg.com). To report issues or for additional feedback, refer to our [Census GitHub repository](https://github.com/chanzuckerberg/cellxgene-census/issues).

Reviewer: also, any links to notebooks, doc pages, etc?

Reviewer 3: Good point, will add some to the introduction.

Reviewer 3: Actually we don't have specific notebooks or docs around these new data additions
